lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs

```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
         F.toList (collateralOutput tx)
     | otherwise =
         outputs tx
+
+--------------------------------------------------------------------------------
+-- Generating and shrinking arbitrary sequences of blocks
+--------------------------------------------------------------------------------
+
+-- | A sequence of blocks that starts at a given block height and slot number.
+--
+-- Values of this type model a sequence of state transitions affecting the
+-- global UTxO set:
+--
+-- @
+--    global_utxo_0 -> block_0_1 ->
+--    global_utxo_1 -> block_1_2 ->
+--    ...
+--    global_utxo_i -> block_i_j ->
+--    global_utxo_j
+-- @
+--
+-- In turn, each block contains a series of transactions that each affect the
+-- global UTxO set. The effect of applying a single block to the global UTxO
+-- set is equivalent to applying all of the contained transactions in order
+-- to the global UTxO set.
+--
+-- In general, only a subset of transactions within a block sequence will be
+-- relevant to a particular wallet, and only a subset of entries within each
+-- successive iteration of the global UTxO set will be relevant to that wallet.
+--
+-- Values of 'BlockSeq' are more convenient to generate and shrink than values
+-- of 'BlockData', but can easily be transformed into values of 'BlockData' in
+-- order to test properties of the 'applyBlocks' function.
+--
+-- Usage:
+--
+--  1. Use function 'genBlockSeq' to generate a 'BlockSeq' value.
+--  2. Use function 'blockSeqToBlockData' to convert a 'BlockSeq' value to a
+--     value of 'BlockData' suitable for use with the 'applyBlocks' function.
+--  3. Use function 'shrinkBlockSeq' to shrink to a minimal counterexample.
+--
+data BlockSeq = BlockSeq
+    { initialBlockHeight
+        :: Quantity "block" Word32
+    , initialSlotNo
+        :: SlotNo
+    , shrinkableTxSeq
+        :: ShrinkableTxSeq
+    }
+    deriving (Eq, Generic, Show)
+
+instance Arbitrary BlockSeq where
+    arbitrary = genBlockSeq
+    shrink = shrinkBlockSeq
+
+genBlockSeq :: Gen BlockSeq
+genBlockSeq = BlockSeq
+    <$> fmap toEnum (choose (0, 100))
+    <*> fmap toEnum (choose (0, 100))
+    <*> genTxSeq genUTxO genAddress
+
+shrinkBlockSeq :: BlockSeq -> [BlockSeq]
+shrinkBlockSeq = genericRoundRobinShrink
+    <@> shrinkMap toEnum fromEnum
+    <:> shrinkMap toEnum fromEnum
+    <:> shrinkTxSeq
+    <:> Nil
+
+-- | Converts a 'BlockSeq' to a value of 'BlockData' that is suitable for use
+--   with the 'applyBlocks' function.
+--
+-- This function fills every slot with exactly one block, leaving no gaps.
+--
+blockSeqToBlockData :: BlockSeq -> BlockData m addr tx state
+blockSeqToBlockData = List . blockSeqToBlockList
+  where
+    blockSeqToBlockList :: BlockSeq -> NonEmpty Block
+    blockSeqToBlockList blockSeq =
+        NE.fromList $ getZipList $ makeBlock
+            <$> ZipList (enumFrom $ blockSeq & initialBlockHeight)
+            <*> ZipList (enumFrom $ blockSeq & initialSlotNo)
```

Comment:
```suggestion
            <$> ZipList (enumFrom $ initialBlockHeight blockSeq)
            <*> ZipList (enumFrom $ initialSlotNo blockSeq)
```

Reply:
I do prefer the existing style here, as it emphasises the **_commonality_** between the two lines: in both cases we're enumerating from _attributes_ of `blockSeq`, and in both cases, we're enumerating from some `initial` value. If lenses were defined, we would have written:
```hs
            <$> ZipList (enumFrom $ blockSeq ^. #initialBlockHeight)
            <*> ZipList (enumFrom $ blockSeq ^. #initialSlotNo)
```
But I don't feel strongly about it.

Reply:
I see your point, usually I also prefer keeping similar parts on the lhs, and varying parts on rhs. 
In this particular case it leads to a mixture of `$` and `&` operators in the same expression and this makes it hard to parse in terms of associativity:
Example: `a $ b & c` - should it be associated to the left as `(a $ b) & c` or `a $ (b & c)` ?
Would you be willing to compromise on this:
```
            <$> ZipList (enumFrom (blockSeq & initialBlockHeight))
            <*> ZipList (enumFrom (blockSeq & initialSlotNo))
```
?

Reply:
Hi @Unisay 

I've included your suggestion in a78e4215baffe74e50d199e8a64c9772d9bcebb4.